File: src/main/java/algo/parenthesis/LongestValidParentheses_32.java
package algo.parenthesis;

import java.util.Stack;

public class LongestValidParentheses_32 {
  public static int longestValidParentheses(String s) {
    int leftCount = 0, rightCount = 0;
    var stack = new Stack<String>();
    for (var i = 0; i < s.length(); i++) {
      var ch = s.charAt(i);
      if (ch == '(') {
        stack.push("(");
        leftCount++;
      } else if (ch == ')') {
        if (!stack.isEmpty()) {
          var p = ")";
          if (rightCount < leftCount) {
            while (!stack.isEmpty()) {
              var e = stack.pop();
              p = e + p;
              if (e.equals("(")) {
                break;
              }
            }
            rightCount++;
          }
          stack.push(p);
        }
      }
    }

    var maxLen = 0;
    var len = 0;
    while (!stack.isEmpty()) {
      var e = stack.pop();
      if (e.equals("(") || e.equals(")")) {
        len = 0;
        continue;
      }
      len += e.length();
      maxLen = Math.max(maxLen, len);
    }
    return maxLen;
  }

  public static void main(String[] args) {
    System.out.println(longestValidParentheses("(()"));
    System.out.println(longestValidParentheses(")()())"));
    System.out.println(longestValidParentheses("(()()()())"));
    System.out.println(longestValidParentheses(")()())()()("));
    System.out.println(longestValidParentheses(")(())))(())())"));
  }
}

File: src/main/kotlin/algo/parenthesis/LongestValidParentheses_32.kt
package algo.parenthesis

import java.util.Stack

class KtLongestValidParentheses_32 {
    fun longestValidParentheses(s: String): Int {
        val stack = Stack<String>()
        var leftParen = 0; var rightParen = 0
        s.forEach { c ->
            if (c == '(') {
                stack.push("(")
                leftParen++
            } else {
                var str = ")"
                if (rightParen < leftParen) {
                    while (stack.isNotEmpty()) {
                        val e = stack.pop()
                        str = e + str
                        if (e == "(") {
                            break
                        }
                    }
                    rightParen++
                }
                stack.push(str)
            }
        }

        var len = 0
        var longest = 0
        stack.forEach {
            if (it == "(" || it == ")") {
                longest = longest.coerceAtLeast(len)
                len = 0
            } else {
                len += it.length
            }
        }
        longest = longest.coerceAtLeast(len)
        return longest
    }
}

fun main() {
    val valid = KtLongestValidParentheses_32()
    println(valid.longestValidParentheses("(()"))
    println(valid.longestValidParentheses(")()())"))
    println(valid.longestValidParentheses(""))
    println(valid.longestValidParentheses(")()())()()("))
}